yapapi/services.py

```diff
@@ -191,26 +204,159 @@ def _respond_nowait(self, message: Optional[Any], response_to: Optional[ServiceS
+        passed as part of the agreement between the requestor and the provider).
+
+        Therefore, this default implementation performs the minimum required for a VM payload to
+        start responding to `run` commands. If your service requires any additional operations -
+        you'll need to override this method (possibly starting with a call to `super().start()`)
+        to add appropriate preparatory steps.
+
+        In case of runtimes other than VM, `deploy` and/or `start` might be optional or altogether
+        disallowed, plus `start` itself might take some parameters. It is up to the author of the
+        specific `Service` implementation that uses such a payload to adjust this method accordingly
+        based on the requirements for the given runtime/exe-unit type.
+        """
 
         self._ctx.deploy()
         self._ctx.start()
         yield self._ctx.commit()
 
-    async def run(self):
-        """Implement the `running` state of the service."""
+    async def run(self) -> AsyncGenerator[WorkItem, Awaitable[List[events.CommandEvent]]]:
+        """Implement the handler for the `running` state of the service.
+
+        To be overridden by the author of a specific Service class.
+
+        Should contain any operations needed to ensure the continual operation of a service.
+
+        As a handler implementing the [work generator pattern](https://handbook.golem.network/requestor-tutorials/golem-application-fundamentals/hl-api-work-generator-pattern),
+        it's expected to be a generator that yields `WorkItems` (generated using the service's
+        instance of the work context) that are then dispatched to the activity by the engine.
+
+        Results of those batches can then be retrieved by awaiting the values captured from yield
+        statements.
+
+        A clean exit from a handler function triggers the engine to transition the state of the
+        instance to the next stage in service's lifecycle - in this case, to `stopping`.
+
+        Any unhandled exception will cause the instance to be terminated.
+
+        ### Example
+
+        ```
+        async def run(self):
+            while True:
+                self._ctx.run("/golem/run/report", "--stats")  # index 0
+                future_results = yield self._ctx.commit()
+                results = await future_results
+                stats = results[0].stdout.strip()  # retrieve from index 0
+                print(f"stats: {stats}")
+        ```
+
+        ### Default implementation
+
+        Because the nature of the operations required during the "running" state depends directly
+        on the specifics of a given Service and because it's entirely plausible for a service
+        not to require any direct interaction with the exe-unit (runtime) from the requestor's end
+        after the service has been started, the default is to just wait indefinitely without
+        producing any batches.
```

Comment:
```suggestion
        Since the logic performed during the `running` state depends entirely on the specifics of a
        given service, the default is to just wait indefinitely without producing batches. Also, it's
        not uncommon for a service not to require any direct interaction with the exe-unit from the
        requestor's end once the service has been started.
```